### PR TITLE
Update gcm3

### LINKED ISSUE #36: BroadcastReceiver trying to return result during a non-ordered broadcast.
http://stackoverflow.com/questions/23835970/broadcastreceiver-trying-to-return-result-during-a-non-ordered-broadcast-packa

I think this is the line causing it?  

https://github.com/inloop/easygcm/blob/master/easygcm/src/main/java/eu/inloop/easygcm/GcmBroadcastReceiver.java#L27

### PR DESCRIPTION
**Pull request by peterbartos**
Update to new GCM using Play Services 7.5. 
Fix #36 #34 #33 #32

### CODE REVIEW

**Comment by DanielNovak** [comment]
I will take a look at this. Adding also @destil to this PR.

**Comment by DanielNovak** [comment]
Back to you, @peterbartos .

File: easygcm-sample/src/main/java/eu/inloop/easygcm/sample/App.java [hunk 1]
```diff
@@ -4,22 +4,20 @@
...
 
 import eu.inloop.easygcm.GcmListener;
-import eu.inloop.easygcm.WakeLockRelease;
 
 public class App extends Application implements GcmListener {
 
     @Override
-    public void onMessage(String s, Bundle bundle, WakeLockRelease wakeLockRelease) {
-        System.out.println("### message: " + s);
+    public void onMessage(String from, Bundle data) {
```
Comment (destil) [inline easygcm-sample/src/main/java/eu/inloop/easygcm/sample/App.java]: Why no wakelocks? Library user might want do some work with wakelock.

**Comment by destil** [comment]
Isn't the google-services.json file with configuration required? Even if it's not, I think it's a nice way how to configure Google APIs. Nothing in resources (config.xml), just one json in the app. And I imagine that the app will need multiple Google APIs, then the app needs just one json for all APIs (login, Drive, GCM, ...) and it will just work.

File: easygcm-sample/src/main/java/eu/inloop/easygcm/sample/App.java (hunk 1 above)
Comment (DanielNovak) [inline easygcm-sample/src/main/java/eu/inloop/easygcm/sample/App.java]: +1. But the approach will be a bit different on GCM 3.0. The GcmReceiver is part of the GCM library and this is starting a wakeful service - which is implemented as EasyGcmListenerService. And it seems the implementation from Google is calling GcmReceiver.completeWakefulIntent() right inside the onStartCommand super implementation - which is kind of strange - I don't see a direct way how we can release the lock ourselves... ?

Comment (DanielNovak) [inline easygcm-sample/src/main/java/eu/inloop/easygcm/sample/App.java]: So I spoke with @destil on this issue and we actually don't have to release the wakelock, it's done by the super class. The important thing is, that each onMessage is executed on a separate thread (spawned by AsyncTask.THREAD_POOL_EXECUTOR). 
We have to update the javadocs and documentation to make sure that the developers will know that onMessage is an asynchronous method.

File: easygcm/src/main/java/eu/inloop/easygcm/GcmHelper.java [hunk 2]
```diff
@@ -185,30 +186,15 @@ void registerInBackground(final Context context, final RegistrationListener regi
         final AsyncTask<Void, Void, Void> registrationTask = new AsyncTask<Void, Void, Void>() {
```
Comment (destil) [inline easygcm/src/main/java/eu/inloop/easygcm/GcmHelper.java]: Look at the demo app here:
git clone https://github.com/googlesamples/google-services.git

They are using IntentService for performing the registrations. I think it's a good practice. Have a look at synchronization as well, the library should prevent multiple registrations at once.

File: easygcm-sample/src/main/java/eu/inloop/easygcm/sample/App.java (hunk 1 above)
Comment (destil) [inline easygcm-sample/src/main/java/eu/inloop/easygcm/sample/App.java]: As we found in documentation, the public void onMessageReceived(String from, Bundle data) method is run in an asynctask and it's ensured that Android will not kill it while AsyncTask is running. The library should state that the onMessage callback will be called on non-ui thread and should not launch new AsyncTasks, it should call networks etc. directly.

**Comment by DanielNovak** [comment]
+1 on the json configuration files. To keep backwards compatibility we can check if easygcm_sender_id string and only use it in that case. Otherwise the json file will be used. The developers will be instructed by the README to put ``` apply plugin: 'com.google.gms.google-services ``` into their build.gradle files.

File: easygcm/src/main/java/eu/inloop/easygcm/GcmHelper.java (hunk 2 above)
Comment (DanielNovak) [inline easygcm/src/main/java/eu/inloop/easygcm/GcmHelper.java]: Some simple synchronization is in place, there is a mRegistrationRunning atomicboolean, and next registrations will be skipped if there is already an registration running. Synchronization is not needed maybe - we should just make sure that we don't run more at once.

Comment (destil) [inline easygcm/src/main/java/eu/inloop/easygcm/GcmHelper.java]: Good point, the current synchronization is enough. How about replacing AsyncTask with an IntentService as in the sample? I think it's nicer and less likely to be killed.

File: easygcm/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -18,30 +19,28 @@
...
-            android:name="eu.inloop.easygcm.GcmBroadcastReceiver"
-            android:permission="com.google.android.c2dm.permission.SEND" >
+            android:name="com.google.android.gms.gcm.GcmReceiver"
+            android:exported="true"
+            android:permission="com.google.android.c2dm.permission.SEND">
             <intent-filter>
                 <!-- Receives the actual messages. -->
                 <action android:name="com.google.android.c2dm.intent.RECEIVE" />
-                <category android:name="${localApplicationId}" />
+                <category android:name="${applicationId}" />
```
Comment (destil) [inline easygcm/src/main/AndroidManifest.xml]: If you want to support pre-4.4 KitKat devices, add the following action to the intent filter declaration for the receiver: <action android:name="com.google.android.c2dm.intent.REGISTRATION" />

**Comment by DanielNovak** [comment]
I am OK with this. 
The AsyncTask can be converted to an wakeful IntentService in a separate pull request.
@destil do you see any more issues here? Thanks.

**Comment by peterbartos** [comment]
Thank you destil for the good review. Good idea with adding the string only as a value without using the plugin. Update soon...

**Comment by destil** [comment]
Code-review +2, Looks good to me

**Comment by DanielNovak** [comment]
Okay, this is ready to merge @peterbartos .

**Pull request merged**